A different approach to a fghack program

g-pape opened the pull request:
chpst: new option -F: fghack using unshare()

chpst with option -F runs prog in new mount and pid namespaces as pid 2. It provides a pid 1 that waits and reaps detached processes until the system reports no more child processes in the pid namespace, and then exits.  While running, chpst -F relays most signals it receives to the pid 1 it provides, which in turn relays the signal to all children in the pid namespace. Except for the quit signal, which tells chpst -F to send the kill signal to the pid 1 it provides to terminate it and all its children immediately.

There's also a new -I option. Nothing that you cannot do with the util-linux unshare(1) program, but comes with the -F option at nearly no cost. Can be used to test supervising another init or running two different init programs side by side.

g-pape committed:
* Makefile, TARGETS: add test for unshare() system call (_GNU_SOURCE).

* hasunshare.h1, hasunshare.h2, tryunshare.c: new.

g-pape committed:
chpst.c: add new option -F: fghack using unshare().

chpst with option -F runs prog in new mount and pid namespaces as pid 2.
It provides a pid 1 that whenever prog or a detached child of prog exits
checks /proc whether it's alone or there's still prog or children of it
running. If it finds itself alone in the pid namespace, it exits. While
running, chpst -F relays signals it receives to the pid 1 it provides,
which in turn relays the signal to all children in the pid namespace.

g-pape committed:
chpst.c: add new option -I: run another init using unshare()

chpst with option -I runs prog in new mount and pid namespaces as pid 1.
prog must be an init program. This can be used to supervise another init
or run two init programs side by side. Use with caution.

g-pape committed:
chpst.c: -F, -I: on QUIT signal send KILL to new pid 1

g-pape committed:
chpst.c: -F: pid1: do not relay CHILD signal to children

g-pape committed:
chpst.c: -F, -I: set mount propagation for / recursively, umount /proc

g-pape committed:
chpst.c: -F: pid1: only check /proc after getting ECHILD

Instead of looking in /proc whenever a child exits, only check /proc
after waitpid() reports ECHILD, which basically means new pid 1 is now
alone. If there's still other processes listed in /proc though, sleep
a few seconds before exiting and so forcefully killing those ghosts.

g-pape committed:
chpst.c: -F: pid1: do not mount and check /proc at all

Be rude and let ghost processes be immediately killed on exit.

capezotte commented:
What if `-I` created a new pid+mount namespace, with `prog...` running as PID 1, and `-F` enabled the mini-pid1 (with `-F` would implying `-I`)?

Currently, `-I` by itself is a no-op.

andy-bower commented:
Hi @g-pape,

This is a nice idea! I was inspired to follow the PID namespace approach to [implement](https://gitlab.com/init-tools/xchpst/-/merge_requests/23) `-F` in my tool, _xchpst_, which aims to be a compatible superset of _chpst_. I also aliased `-I` to xchpst's `--pid-ns` option (and adjusted the mini supervisor to translate `SIGQUIT` as your implementation does). This leads me to make a suggestion about how the new options are defined:

Since `-F` has a superset of the behaviour of `-I` it would be nice if `-FI` behaved the same as `-F` rather than `-I`. That would make the options more composable and also open up the possibility of a future alternative implementation of `-F` not using a PID namespace. And in that case you could recommend in the documentation that users who specifically want a PID namespace not just as a side effect of the fghack behaviour should specify both options explicitly.

Would it be possible to adjust the logic a bit so that it was more like this?

| options | current behaviour | proposed behaviour|
|-----|-----|-----|
| `-I` | pidns | pidns |
| `-F` | pidns + fghack| pidns + fghack | 
| `-FI` | pidns | pidns + fghack |

I guess to do this you would swap the order of the case statements and change how the two boolean config variables work.

I hope this is helpful feedback!

g-pape committed:
chpst.c: cleanup; usage; adjust #ifdefs, error messages

g-pape committed:
chpst.c: -F option implies -I (thx github/andy-bower)

Since -F has a superset of the behaviour of -I, -FI now behaves the same
as -F rather than -I.

g-pape commented:
Yes, this makes sense. With commit 6875024 the -F option implies -I, and if both options are given, -F now takes precedence. Please have a look!

andy-bower commented:
> Yes, this makes sense. With commit [6875024](https://github.com/g-pape/runit/commit/68750248985ab210503bb8dbe7aefc5813476696) the -F option implies -I, and if both options are given, -F now takes precedence. Please have a look!

Logic looks good, thanks! I haven't tried it out yet.

andy-bower commented:
> > Yes, this makes sense. With commit [6875024](https://github.com/g-pape/runit/commit/68750248985ab210503bb8dbe7aefc5813476696) the -F option implies -I, and if both options are given, -F now takes precedence. Please have a look!
> 
> Logic looks good, thanks! I haven't tried it out yet.

Tried a few things and it seems to work as expected.

With `fghack` there seem to be so many questions where there isn't a definite right answer, such as what return code should prevail in what situation.

g-pape commented:
chpst generally exits with the return code of `prog`. Since with option -F `prog` can be many processes, this chpst version exits with the return code of the last process that exited in the namespace.

g-pape committed:
Merge branch 'next' into topic/chpst-add-options-F-and-I

g-pape committed:
md/chpst.8.md: document new -F and -I options
